### PR TITLE
Update to handle JSON-RPC amounts in either the old or new conventions

### PR DESCRIPTION
**Pull request by ksedgwic**
The JSON-RPC interface has changed the type of amounts from string w/ "msat" suffix to a plain number.

The backtrace hack and removal of the Amount constructor-from-string are useful during debugging but we might want to drop them ...

### CODE REVIEW

**Comment by tsjk** [comment]
What does this fix?
Does it make clboss compatible with cln v23?

**Comment by ksedgwic** [comment]
Yes, this PR makes CLBOSS compatible with `v23.05`.

However, it is not currently backwards compatible; if you run this PR against older versions of CLN it won't work.  It should be possible to make it backwards compatible, I will have a look at that next ...

**Comment by ksedgwic** [comment]
@tsjk I've reworked the CLBOSS amount PR so it should be able to handle either the old or new conventions.

**Comment by ksedgwic** [comment]
It's running successfully on CLN v23.05, but I have not verified it on an older CLN.

**Comment by tsjk** [comment]
I'll try to find time to test this soon.

**Comment by ksedgwic** [comment]
@tsjk switching back to Draft, I'm seeing bogus swap reports in `clboss-status`:
```
         {
            "time": 1.68391e+09,
            "time_human": "UTC 2023-05-12 17:29:15.552",
            "amount_sent": "2msat",
            "amount_received": "18446744073709551456msat",
            "amount_lost": "0msat",
            "provider": "Boltz::Service(\"https://testnet.boltz.exchange/api\")"
         },
```
The `amount_sent` and `amount_received` are obviously wrong.

The values in the `data.clboss` db are also bad after my changes:
```
1679187607.98172|299528669|297715000|Boltz::Service("https://testnet.boltz.exchange/api")
1679200208.08367|100094092|99148000|Boltz::Service("https://testnet.boltz.exchange/api")
1679211008.25419|398096693|395157000|Boltz::Service("https://testnet.boltz.exchange/api")
1679291804.75702|108338105|105921000|Boltz::Service("https://testnet.boltz.exchange/api")
1679302004.89865|100000095|92722000|Boltz::Service("https://testnet.boltz.exchange/api")
1680386607.52125|4096999375|4077496000|Boltz::Service("https://testnet.boltz.exchange/api")
1681898663.87814|312002872|309660000|Boltz::Service("https://testnet.boltz.exchange/api")
1683855569.11341|2|94067341782384|Boltz::Service("https://testnet.boltz.exchange/api")
1683867701.28066|2|94842369897008|Boltz::Service("https://testnet.boltz.exchange/api")
1683878085.26563|2|-160|Boltz::Service("https://testnet.boltz.exchange/api")
1683885285.61652|2|-160|Boltz::Service("https://testnet.boltz.exchange/api")
1683893817.04882|2|-160|Boltz::Service("https://testnet.boltz.exchange/api")
1683902355.21184|2|-160|Boltz::Service("https://testnet.boltz.exchange/api")
1683912555.55297|2|-160|Boltz::Service("https://testnet.boltz.exchange/api")
```

Still looking, clues welcome ...

**Comment by tsjk** [comment]
The negative numbers don't look too good. :)
I have not detailed knowledge of the code, but I have browsed through it on several occasions.
The amount received is on-chain, right? Are on-chain transactions effected in the same way by the api-change in cln?
If you compare what the numbers should be with what they are - does that give clues? How much did you actually send and receive?

**Comment by ksedgwic** [comment]
@tsjk I force pushed a new version which uses a class method instead of a conversion constructor for the JSON -> Amount method.  I have seen one swap since and it was correctly reported.  Possibly the problem is fixed.

TODO:
- [x] observe additional swap reports with v23.05
- [ ] test on older versions which still use deprecated amount msat string format.

**Comment by ksedgwic** [comment]
@tsjk I've observed more swaps with well-formed reports.  I'm checking the first TODO box.  Any chance you could try on an older CLN version to validate backwards compatibility?

**Comment by tsjk** [comment]
Sorry. I've been busy with other things.
I guess I could test this. Haven't ran cln on a testnet but can imagine to try it and let clboss go wild.
Not sure testing it on my mainnet node would make a lot of sense.

**Comment by JaviLib** [comment]
When is the solution going to be merged with master? Is it ok to wait?

**Comment by ksedgwic** [comment]
It still needs testing on older systems with the legacy msat format, modulo that it has been working great for us on `v23.05`

**Comment by tsjk** [comment]
I ended up upgrading my mainnet node, because the version I was using (v22.11.1 with some back-ported patches) actually back-ported the msat removal - voiding my reason for stay on that version. For now my build recipe for Gentoo only applies this patch if the installed version of core lightning is v23.05 or higher.
But, we should definitely try to test this. My idea was to setup at testnet node with a testnet bitcoind. I realise that's quite an effort. So, perhaps this should be tested by someone still on earlier version - or we could just go high chaparral and requiring that the latest clboss requires core-lightning v23.05 or later...

**Comment by ksedgwic** [comment]
> or we could just go high chaparral and requiring that the latest clboss requires core-lightning v23.05 or later...

Over time this will become more and more reasonable.  Not sure if it is quite reasonable yet though (especially for mainnet where folks run more mature versions ...)

**Comment by JaviLib** [comment]
Well, people using old versions of Core Lightning and not intending to upgrade have zero reasons to try this patch. So I think it is better to just release this and pass an important notice like "Not compatible with versions previous to v23.05". You may want to even test version in code and return an error if version is not equal or higher than v23.05..

**Comment by tsjk** [comment]
Yes, checking for version sounds sensible.
On another note, I've been running this for a while and it seems ok. No submarine swaps performed as of yet, however.

**Comment by JaviLib** [comment]
Would you please check if #147 is related to this?

**Comment by chrisguida** [comment]
Testing this on 23.02 with a raspiblitz

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ZmnSCPxj/clboss/pull/162/commits/a4bb0192550803db3d07628a29284a76f7204365

**Comment by chrisguida** [comment]
Update: ran overnight on raspiblitz with CLN 23.02, looks great. Let's merge!

ACK a4bb0192550803db3d07628a29284a76f7204365

**Review by chrisguida — APPROVED** [review]

**Comment by ksedgwic** [comment]
I'm attempting resolve a conflict with recent commits on `master`.

Having a build problem with master before any rebasing/merging: #166

**Comment by ksedgwic** [comment]
rebasing on master

**Comment by ksedgwic** [comment]
I'm going to smoke test this in my test nodes, will merge if problem free

**Comment by tsjk** [comment]
Was just about to ask whether this was tested on any pre-msat-drop nodes in the end.

**Comment by ksedgwic** [comment]
@tsjk https://github.com/ZmnSCPxj/clboss/pull/162#issuecomment-1682803646

**Comment by ksedgwic** [comment]
Looks stable on my testnet nodes, merging

**Comment by kn0wmad** [comment]
Any ETA on a release that might include this fix?

**Comment by ksedgwic** [comment]
@kn0wmad https://github.com/ZmnSCPxj/clboss/issues/172

**Pull request merged**
